sycl/include/sycl/ext/oneapi/annotated_arg/annotated_arg.hpp

```diff
@@ -0,0 +1,301 @@
+//==----------- annotated_arg.hpp - SYCL annotated_arg extension -----------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <cstddef>
+#include <type_traits>
+
+#include <sycl/detail/stl_type_traits.hpp>
+#include <sycl/exception.hpp>
+#include <sycl/ext/oneapi/annotated_arg/properties.hpp>
+#include <sycl/ext/oneapi/properties/properties.hpp>
+
+#ifdef __SYCL_DEVICE_ONLY__
+#define __SYCL_HOST_NOT_SUPPORTED(Op)
+#else
+#define __SYCL_HOST_NOT_SUPPORTED(Op)                                          \
+  throw sycl::exception(                                                       \
+      sycl::make_error_code(sycl::errc::feature_not_supported),                \
+      Op " is not supported on host device.");
+#endif
+
+#ifdef __SYCL_DEVICE_ONLY__
+#define __OPENCL_GLOBAL_AS__ __attribute__((opencl_global))
+#ifdef __ENABLE_USM_ADDR_SPACE__
+#define __OPENCL_GLOBAL_DEVICE_AS__ __attribute__((opencl_global_device))
+#define __OPENCL_GLOBAL_HOST_AS__ __attribute__((opencl_global_host))
+#else
+#define __OPENCL_GLOBAL_DEVICE_AS__ __attribute__((opencl_global))
+#define __OPENCL_GLOBAL_HOST_AS__ __attribute__((opencl_global))
+#endif // __ENABLE_USM_ADDR_SPACE__
+#define __OPENCL_LOCAL_AS__ __attribute__((opencl_local))
+#define __OPENCL_CONSTANT_AS__ __attribute__((opencl_constant))
+#define __OPENCL_PRIVATE_AS__ __attribute__((opencl_private))
+#else
+#define __OPENCL_GLOBAL_AS__
+#define __OPENCL_GLOBAL_DEVICE_AS__
+#define __OPENCL_GLOBAL_HOST_AS__
+#define __OPENCL_LOCAL_AS__
+#define __OPENCL_CONSTANT_AS__
+#define __OPENCL_PRIVATE_AS__
+#endif
+
+namespace sycl {
+__SYCL_INLINE_VER_NAMESPACE(_V1) {
+namespace ext {
+namespace oneapi {
+namespace experimental {
+
+namespace detail {
+
+// Type-trait for checking if a type defines `operator[]`.
+template <typename T, typename = void>
+struct HasSubscriptOperator : std::false_type {};
+
+template <typename T>
+struct HasSubscriptOperator<
+    T, typename std::enable_if_t<
+           !std::is_void<decltype(std::declval<T>().operator[](0))>::value>>
+    : std::true_type{};
```

Comment:
Just a small simplification:
```suggestion
template <typename T>
struct HasSubscriptOperator : std::bool_constant<!std::is_void<decltype(std::declval<T>().operator[](0))>::value> {};
```